scale_build/packages/package.py

```diff
@@ -0,0 +1,189 @@
+
+        return self._binary_packages
+
+    def build_time_dependencies(self, all_binary_packages=None):
+        if self._build_time_dependencies is not None:
+            return self._build_time_dependencies
+        elif not all_binary_packages:
+            raise CallError('Binary packages must be specified when computing build time dependencies')
+
+        self._build_time_dependencies = get_install_deps(
+            all_binary_packages, set(), self.build_depends
+        ) | self.explicit_deps
+        return self._build_time_dependencies
+
+    @property
+    def hash_changed(self):
+        if self.name == 'truenas':
+            # truenas is special and we want to rebuild it always
+            # TODO: Do see why that is so
+            return True
+
+        source_hash = self.source_hash
+        existing_hash = None
+        if os.path.exists(self.hash_path):
+            with open(self.hash_path, 'r') as f:
+                existing_hash = f.read().strip()
+        if source_hash == existing_hash:
+            return run(
+                ['git', '-C', self.source_path, 'diff-files', '--quiet', '--ignore-submodules'], check=False
+            ).returncode != 0
+        else:
+            return True
+
+    @property
+    def source_hash(self):
+        return run(['git', '-C', self.source_path, 'rev-parse', '--verify', 'HEAD']).stdout.decode().strip()
+
+    @property
+    def rebuild(self):
+        return self.hash_changed or self.parent_changed
+
+    @property
+    def hash_path(self):
+        return os.path.join(HASH_DIR, f'{self.name}.hash')
+
+    @property
+    def exists(self):
+        return os.path.exists(self.source_path)
+
+    def retrieve_current_remote_origin_and_sha(self):
+        if self.exists:
+            return retrieve_git_remote_and_sha(self.source_path)
+        else:
+            return {'url': None, 'sha': None}
+
+    def update_git_manifest(self):
+        info = self.retrieve_current_remote_origin_and_sha()
+        update_git_manifest(info['url'], info['sha'])
+
+    def checkout(self, branch_override=None):
+        origin_url = self.retrieve_current_remote_origin_and_sha()['url']
+        branch = branch_override or self.branch
+        git_logger = get_logger(f'checkout_{self.name}', 'git-checkout.log', 'w')
+        if branch == self.existing_branch and self.origin == origin_url:
+            logger.debug('Updating git repo [%s] (%s)', self.name, GIT_LOG_PATH)
+            run(['git', '-C', self.source_path, 'fetch', '--unshallow'], logger=git_logger, check=False)
+            run(['git', '-C', self.source_path, 'fetch', 'origin', branch], logger=git_logger)
+            run(['git', '-C', self.source_path, 'reset', '--hard', f'origin/{branch}'], logger=git_logger)
+        else:
+            logger.debug('Checking out git repo [%s] (%s)', self.name, GIT_LOG_PATH)
+            shutil.rmtree(self.source_path, ignore_errors=True)
```

Comment:
What errors are you ignoring?

We expect `self.source_path` to be non-existent after this method terminates successfully. Will it always be so?